app/repositories/RepositoryOps.scala

```diff
@@ -0,0 +1,103 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package repositories
+
+import com.mongodb.ErrorCategory.DUPLICATE_KEY
+import com.mongodb.client.model.{ReturnDocument, Updates}
+import org.mongodb.scala.bson.BsonDocument
+import org.mongodb.scala.model.Filters.equal
+import org.mongodb.scala.model.{FindOneAndReplaceOptions, FindOneAndUpdateOptions}
+import org.mongodb.scala.{Document, MongoCollection, MongoWriteException}
+import play.api.Logging
+import play.api.libs.json.{Json, Writes}
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.reflect.ClassTag
+
+trait RepositoryOps[T] extends Logging {
+
+  implicit def classTag: ClassTag[T]
+  implicit val executionContext: ExecutionContext
+
+  val collection: MongoCollection[T]
+
+  def findAll: Future[Seq[T]] =
+    collection.find().toFuture
+
+  def findAll[V](keyId: String, keyValue: V): Future[Seq[T]] =
+    collection.find(equal(keyId, keyValue)).toFuture
+
+  def findOne[V](keyId: String, keyValue: V): Future[Option[T]] =
+    collection.find(equal(keyId, keyValue)).toFuture.map(_.headOption)
+
+  /*
+   Find one and return if a document with keyId=keyValue exists,
+   or create "document: T" if a document with keyId=keyValue does NOT exists.
+   */
+  def findOneOrCreate[V](keyId: String, keyValue: V, document: => T)(implicit writes: Writes[T]): Future[T] =
+    collection
+      .findOneAndUpdate(
+        filter = equal(keyId, keyValue),
+        update = Updates.setOnInsert(BsonDocument(Json.toJson(document).toString)),
+        options = FindOneAndUpdateOptions().upsert(true).returnDocument(ReturnDocument.AFTER)
+      )
+      .toFuture
+
+  /*
+   Find one and replace with "document: T" if a document with keyId=keyValue exists,
+   or create "document: T" if a document with keyId=keyValue does NOT exists.
+   */
+  def findOneAndReplace[V](keyId: String, keyValue: V, document: T): Future[T] =
+    collection
+      .findOneAndReplace(
+        filter = equal(keyId, keyValue),
+        replacement = document,
+        options = FindOneAndReplaceOptions().upsert(true).returnDocument(ReturnDocument.AFTER)
+      )
+      .toFuture
+
+  def findOneAndRemove[V](keyId: String, keyValue: V): Future[Option[T]] =
+    collection.findOneAndDelete(equal(keyId, keyValue)).toFutureOption
+
+  def indexList: Future[Seq[Document]] = collection.listIndexes.toFuture
+
+  def insertTest(document: T): Future[Either[MongoError, T]] = throw new RuntimeException("test")
+
+  def insertOne(document: T): Future[Either[MongoError, T]] =
+    collection
+      .insertOne(document)
+      .toFuture
+      .map(_ => Right(document))
+      .recover {
+        case exc: MongoWriteException if (exc.getError.getCategory == DUPLICATE_KEY) => Left(DuplicateKey)
+      }
```

Comment:
Would it be more useful for the return type to be `Future[Either[MongoWriteException, T]]` rather than replacing the exception with our `DuplicateKey` case object (I'm assuming the exception might have more details about the key that already exists)?